src/datasets/ground_atom_data.py

```
@@ -24,31 +25,35 @@ def create_ground_atom_data(env: BaseEnv, base_dataset: Dataset,
-                    pred = ga.predicate
-                    if (totals[pred] == 0 or \
-                            labeleds[pred] / totals[pred] <= annotating_ratio):
-                        # Teacher comments on this atom
-                        stripped_pred = predicates_to_stripped[pred]
-                        stripped_ga = GroundAtom(stripped_pred, ga.objects)
-                        subset_atoms.add(stripped_ga)
-                        labeleds[pred] += 1
-                        annotated_preds.add(pred)
-                    totals[pred] += 1
-            ground_atoms_traj.append(subset_atoms)
-        assert len(traj.states) == len(ground_atoms_traj)
-        annotations.append(ground_atoms_traj)
-    assert len(annotations) == len(base_dataset.trajectories)
-    assert annotated_preds == annotating_predicates, \
-        "Not all predicates are seen in the dataset"
+    # Generate all positive and negative examples
+    pos_examples = DefaultDict(list)  # predicate: [(i, j, ground_atom)]
+    neg_examples = DefaultDict(list)
+    for i, traj in enumerate(base_dataset.trajectories):
+        for j, s in enumerate(traj.states):
+            ground_atoms = utils.abstract(s, annotating_predicates)
+            ground_atom_universe = utils.all_possible_ground_atoms(
+                s, annotating_predicates)
+            for ga in ground_atom_universe:
+                if ga in ground_atoms:  # positive example
+                    pos_examples[ga.predicate].append((i, j, ga))
+                else:
+                    neg_examples[ga.predicate].append((i, j, ga))
+    # Sample `num_examples` from each list
+    pos_picks: List[Tuple] = []
+    neg_picks: List[Tuple] = []
+    for p in annotating_predicates:
+        for examples, picks in zip((pos_examples, neg_examples),
+                                   (pos_picks, neg_picks)):
+            idxs = np.random.choice(len(examples[p]),
+                                    size=num_examples,
```

can you raise an informative error when `num_examples > len(examples[p])` and add a test case for this?

---

is this ok?

---

Yes perf thanks